common/src/BasePlugin.php

```diff
@@ -218,4 +242,85 @@ private function isUniqueAttribute(array $array, string $value): bool {
     private function isValidGtin(string $value): bool {
         return preg_match('/^\d{8}(?:\d{4,6})?$/', $value);
     }
+
+    public function showCustomNotice() {
+        $class = 'notice notice-info is-dismissible ' . $this->getCustomNoticeClass();
+        $message = $this->getCustomNoticeText();
+        if (!empty($message)) {
+            printf('<div class="%s"><p>%s</p></div>', esc_attr($class), $message);
+        }
+    }
+
+    public function getCustomNoticeText() {
+        $readme = wp_remote_fopen($this->getPluginReadme());
+        $pattern = '/===\sUpgrade\sNotice\s===\s* 
+               =\s' . preg_quote($this->getPluginVersion($this->getSlug())) . '\s=\s
+               (?:.+\R)?(.+)/x';
+        preg_match($pattern, $readme, $matches);
+        if (isset($matches[1])) {
+            return $this->convertReadmeLinkToHtml($matches[1]);
+        }
+    }
+
+    private function getPluginReadme(): string {
+        return sprintf(
+            self::PLUGIN_README_URL,
+            $this->getSlug()
+        );
+    }
+
+    private function convertReadmeLinkToHtml(string $notice_text): string {
+        $pattern = '/\[(?<link_label>.+)\]\[(?<link_url>.+)\]/';
+        preg_match($pattern, $notice_text, $link_matches);
+        if (isset($link_matches['link_label'], $link_matches['link_url'])) {
+            $link = sprintf('<a target="_blank" href="%s">%s</a>',
+                $link_matches['link_url'],
+                $link_matches['link_label']
+            );
+            return preg_replace($pattern, $link, $notice_text, 1);
+        }
+        return $notice_text;
+    }
```

Comment:
Why do all of this complicated stuff? Why insist on having the upgrade notices in Markdown? Why don't you just add them to an array in PHP or a separate file, already with HTML? Would save a lot of code...